general/mem_manager.hpp

```diff
@@ -394,158 +434,254 @@ class Memory
    void PrintPtrs(void);
+
+   /// Prints all aliases known by the memory manager
+   void PrintAliases(void);
+
+   MemoryType GetHostMemoryType() { return host_mem_type; }
+   MemoryType GetDeviceMemoryType() { return device_mem_type; }
 };
 
 
 // Inline methods
 
+template <typename T>
+inline void Memory<T>::Reset()
+{
+   h_ptr = NULL;
+   h_mt = MemoryManager::host_mem_type;
+   capacity = 0;
+   flags = 0;
+}
+
+template <typename T>
+inline void Memory<T>::Reset(MemoryType mt)
+{
+   h_ptr = NULL;
+   h_mt = mt;
+   capacity = 0;
+   flags = 0;
+}
+
+template <typename T>
+inline void Memory<T>::New(int size)
+{
+   capacity = size;
+   flags = OWNS_HOST | VALID_HOST;
+   h_mt = MemoryManager::host_mem_type;
+   h_ptr = (h_mt == MemoryType::HOST) ? new T[size] :
+           (T*)MemoryManager::New_(nullptr, size*sizeof(T), h_mt, flags);
+   MemoryManager::CheckHostMemoryType_(h_mt, h_ptr);
+}
+
 template <typename T>
 inline void Memory<T>::New(int size, MemoryType mt)
 {
-   if (mt == MemoryType::HOST)
-   {
-      New(size);
-   }
-   else
+   capacity = size;
+   if (mt == MemoryType::HOST) { flags = OWNS_HOST | VALID_HOST; }
+   h_mt = IsHostMemory(mt) ? mt : MemoryManager::GetDualMemoryType_(mt);
+   T *h_tmp = (h_mt == MemoryType::HOST) ? new T[size] : nullptr;
+   h_ptr = (mt == MemoryType::HOST) ? h_tmp:
+           (T*)MemoryManager::New_(h_tmp, size*sizeof(T), mt, flags);
+   MemoryManager::CheckHostMemoryType_(h_mt, h_ptr);
+}
+
+template <typename T>
+inline void Memory<T>::Wrap(T *host_ptr, int size, bool own)
+{
+   capacity = size;
+   h_ptr = host_ptr;
+   h_mt = MemoryType::HOST;
+   flags = (own ? OWNS_HOST : 0) | VALID_HOST;
+#ifdef MFEM_DEBUG
+   if (own && MemoryManager::Exists())
    {
-      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
-      // type, e.g. CUDA.
-      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
-      h_ptr = (T*)MemoryManager::New_(tmp, size*sizeof(T), mt, flags);
-      capacity = size;
+      MFEM_VERIFY(!MemoryManager::IsKnown_(h_ptr), "Wrap but registered");
+      MFEM_VERIFY(!MemoryManager::IsAlias_(h_ptr), "Wrap but existing alias");
    }
+#endif
 }
 
 template <typename T>
 inline void Memory<T>::Wrap(T *ptr, int size, MemoryType mt, bool own)
 {
-   if (mt == MemoryType::HOST)
-   {
-      Wrap(ptr, size, own);
-   }
-   else
-   {
-      // Allocate the host pointer with new T[] if 'mt' is a pure device memory
-      // type, e.g. CUDA.
-      T *tmp = (mt == MemoryType::CUDA) ? new T[size] : NULL;
-      h_ptr = (T*)MemoryManager::Register_(ptr, tmp, size*sizeof(T), mt, own,
-                                           false, flags);
-      capacity = size;
-   }
+   h_mt = mt;
+   capacity = size;
+   T *h_tmp = (h_mt == MemoryType::HOST) ? new T[size] : nullptr;
+   h_ptr = (T*)MemoryManager::Register_(ptr, h_tmp, size*sizeof(T),
+                                        mt, own, false, flags);
```

Comment:
When `h_mt == MemoryType::HOST`, we can skip the call to `MemoryManager::Register_`. Or is that no longer the case?

Reply:
I think the issue is that `HOST` memory can be wraped, but might need to be registered.
The case is for example `hypre` data being wrapped, and then needed to be on the device. That requires some registration.

Reply:
In the hypre case you describe, there is still no need to register the pointer if it is `MemoryType::HOST` -- it will be registered when copied/moved to device by the methods `ReadWrite`, `Read`, or `Write`.

Reply:
Rerouted the `Wrap` to the `MemoryType::HOST` one in that case.